internal/datacoord/handler.go

```diff
@@ -45,49 +46,84 @@ func newServerHandler(s *Server) *ServerHandler {
 	return &ServerHandler{s: s}
 }
 
-// GetVChanPositions gets vchannel latest postitions with provided dml channel names
+// GetVChanPositions gets vchannel latest postitions with provided dml channel names,
+// we expect QueryCoord gets the indexed segments to load, so the flushed segments below are actually the indexed segments,
+// the unflushed segments are actually the segments without index, even they are flushed.
 func (h *ServerHandler) GetVChanPositions(channel *channel, partitionID UniqueID) *datapb.VchannelInfo {
 	// cannot use GetSegmentsByChannel since dropped segments are needed here
 	segments := h.s.meta.SelectSegments(func(s *SegmentInfo) bool {
 		return s.InsertChannel == channel.Name
 	})
+	segmentInfos := make(map[int64]*SegmentInfo)
+	indexedSegments := FilterInIndexedSegments(h.s.meta, h.s.indexCoord, segments...)
+	indexed := make(typeutil.UniqueSet)
+	for _, segment := range indexedSegments {
+		indexed.Insert(segment.GetID())
+	}
 	log.Info("GetSegmentsByChannel",
 		zap.Any("collectionID", channel.CollectionID),
 		zap.Any("channel", channel),
 		zap.Any("numOfSegments", len(segments)),
 	)
-	var flushedIds []int64
-	var unflushedIds []int64
-	var droppedIds []int64
-	var seekPosition *internalpb.MsgPosition
+	var (
+		flushedIds   = make(typeutil.UniqueSet)
+		unflushedIds = make(typeutil.UniqueSet)
+		droppedIds   = make(typeutil.UniqueSet)
+		seekPosition *internalpb.MsgPosition
+	)
 	for _, s := range segments {
 		if (partitionID > allPartitionID && s.PartitionID != partitionID) ||
 			(s.GetStartPosition() == nil && s.GetDmlPosition() == nil) {
 			continue
 		}
-
+		segmentInfos[s.GetID()] = s
 		if s.GetState() == commonpb.SegmentState_Dropped {
-			droppedIds = append(droppedIds, trimSegmentInfo(s.SegmentInfo).GetID())
-			continue
+			droppedIds.Insert(s.GetID())
+		} else if indexed.Contain(s.GetID()) {
+			flushedIds.Insert(s.GetID())
+		} else {
+			unflushedIds.Insert(s.GetID())
+		}
+	}
+	for id := range unflushedIds {
+		// Indexed segments are compacted to a raw segment,
+		// replace it with the indexed ones
+		if !indexed.Contain(id) &&
+			len(segmentInfos[id].GetCompactionFrom()) > 0 &&
+			indexed.Contain(segmentInfos[id].GetCompactionFrom()...) {
+			flushedIds.Insert(segmentInfos[id].GetCompactionFrom()...)
+			unflushedIds.Remove(id)
```

Comment:
Is it safe to remove an element during iteration.

Reply:
@sunby The `Set[T]` type is actually `map[T]struct{}`, so it is